File: backend/schemas/bid_wizard.py
"""Pydantic schemas for AI编标（bid-wizard）."""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from backend.agent.bid_draft_agent import OUTLINE_MAX_NODES
from backend.agent.bid_wizard_agent import GENERATION_WORD_COUNT_MAX, GENERATION_WORD_COUNT_MIN

WizardStage = Literal["material", "requirement", "outline", "writing"]
GenerationPart = Literal["business", "technical"]

# 待答问题上限：round 闸门按"未作答/待确认(supplemented)"计数——已答/已采纳/已跳过
# 的问题不算待办、不阻断再次检查（2026-09-11 反馈⑯：20 题全已采纳仍被总数顶死）。
QUESTIONNAIRE_MAX_QUESTIONS = 20
# 跨轮累计问题总数上限：已答问题保留在卡片列表（可改答），轮次多了总数会涨，
# 保存载荷与追加写入按此上限校验。
QUESTIONNAIRE_TOTAL_MAX_QUESTIONS = 60
SPEC_INSTRUCTION_MAX_CHARS = 2_000


# ------------------------------------------------------------------ access


class WizardAccessResponse(BaseModel):
    enabled: bool
    mode: str


# ------------------------------------------------------------------ wizard


class WizardCreate(BaseModel):
    project_id: str | None = Field(default=None, min_length=1, max_length=36)
    project_name: str | None = Field(default=None, max_length=200)


class WizardResponse(BaseModel):
    id: str
    project_id: str
    stage: str
    status: str
    analysis: dict[str, Any] | None
    questionnaire: dict[str, Any] | None
    requirements: dict[str, Any] | None
    spec: list[Any] | None
    spec_previous: list[Any] | None
    spec_confirmed_at: datetime | None
    requirements_stale: bool
    spec_stale: bool
    error_message: str | None
    # 「额外的需求」问答历史（决策 43）：时序正序，条目 {question, answer, created_at, adopted}
    qa_history: list[Any] | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class WizardStageUpdate(BaseModel):
    stage: WizardStage


class WritingTaskBrief(BaseModel):
    """列表页「撰写中」徽标用的最新撰写任务摘要（决策 25）。"""

    id: str
    status: str

    model_config = ConfigDict(from_attributes=True)


class WizardListItem(BaseModel):
    """项目列表页单行（§4.0）：向导 + 项目名 + 招标文件名 + 撰写状态聚合。"""

    wizard_id: str
    project_id: str
    project_name: str
    stage: str
    status: str
    tender_filename: str | None
    latest_writing_task: WritingTaskBrief | None
    updated_at: datetime
    created_at: datetime


class WizardListResponse(BaseModel):
    active: list[WizardListItem]
    archived: list[WizardListItem]


# ------------------------------------------------------------------ materials


class WizardMaterialResponse(BaseModel):
    id: str
    document_id: str
    category: str | None
    index_status: str
    indexed_at: datetime | None
    index_error: str | None
    chunk_count: int | None
    # snapshot of the underlying document for display
    original_filename: str | None
    doc_status: str | None
    word_count: int | None
    page_count: int | None
    file_size: int | None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class WizardMaterialIndexResponse(BaseModel):
    material_id: str
    index_status: str
    chunk_count: int | None
    index_content: str | None


class WizardEstimateResponse(BaseModel):
    """上传前预估：token 量级 + 约点数（null=价目缺失，前端退回 token 提示）。"""

    chars: int
    estimated_tokens: int
    estimated_points: int | None = None


# ------------------------------------------------------------------ questionnaire / requirements


class QuestionnaireAnswer(BaseModel):
    question_id: str = Field(min_length=1, max_length=100)
    # answered（自定义回答）/ adopted（采纳建议答案）/ skipped /
    # supplemented（该问题已补充新素材关闭，答案待下一轮 AI 检查给出）
    action: Literal["answered", "adopted", "skipped", "supplemented"]
    answer: str | None = Field(default=None, max_length=4_000)


class GenerationOptions(BaseModel):
    """生成要求（决策 38-41）：编写需求里固定的三项硬约束子结构。

    word_count 允许为空以保存草稿；必填校验在进入「编写大纲」时由 stage 端点做（决策 40/42）。
    """

    parts: list[GenerationPart] = Field(default_factory=lambda: ["technical"], min_length=1)
    word_count: int | None = Field(
        default=None, ge=GENERATION_WORD_COUNT_MIN, le=GENERATION_WORD_COUNT_MAX
    )
    charts: bool = True

    @field_validator("parts")
    @classmethod
    def _dedupe_parts(cls, value: list[str]) -> list[str]:
        deduped: list[str] = []
        for item in value:
            if item not in deduped:
                deduped.append(item)
        return deduped


class RequirementsUpdate(BaseModel):
    answers: list[QuestionnaireAnswer] = Field(min_length=0, max_length=QUESTIONNAIRE_TOTAL_MAX_QUESTIONS)
    # 不带该字段=沿用已保存的生成要求（老客户端 / 只改作答）；带则整体覆盖
    generation_options: GenerationOptions | None = None


class WizardQaAsk(BaseModel):
    """追问侧栏提问（决策 32a）。"""

    question: str = Field(min_length=1, max_length=2_000)


class WizardQaAskResponse(BaseModel):
    answer: str


class WizardQaAdopt(BaseModel):
    """采纳追问问答对并入编写需求（supplementals）。"""

    question: str = Field(min_length=1, max_length=2_000)
    answer: str = Field(min_length=1, max_length=4_000)


# ------------------------------------------------------------------ spec


class SpecChartPlan(BaseModel):
    type: Literal["table", "mermaid"] = "table"
    title: str = Field(min_length=1, max_length=200)
    points: str | None = Field(default=None, max_length=1_000)


class SpecNode(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    level: int = Field(default=1, ge=1, le=6)
    summary: str | None = Field(default=None, max_length=2_000)
    article_count: int = Field(default=2, ge=1, le=8)
    text_count: int = Field(default=400, ge=100, le=3_000)
    charts: list[SpecChartPlan] | None = Field(default=None, max_length=6)


class SpecUpdate(BaseModel):
    spec: list[SpecNode] = Field(min_length=1, max_length=OUTLINE_MAX_NODES)


class SpecRevise(BaseModel):
    instruction: str = Field(min_length=1, max_length=SPEC_INSTRUCTION_MAX_CHARS)


# ------------------------------------------------------------------ writing tasks


class WritingTaskCreate(BaseModel):
    node_ids: list[str] = Field(min_length=1, max_length=OUTLINE_MAX_NODES)


class WritingTaskResponse(BaseModel):
    id: str
    wizard_id: str
    project_id: str
    status: str
    selected_nodes: list[Any] | None
    summary: dict[str, Any] | None
    continue_of: str | None
    error_message: str | None
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class WizardSectionResponse(BaseModel):
    node_id: str
    title: str
    summary: str | None
    chart_plan: list[Any] | None
    status: str
    word_count: int | None
    written_at: datetime | None
    attempts: int
    error_message: str | None

    model_config = ConfigDict(from_attributes=True)


class WizardSectionContentResponse(BaseModel):
    node_id: str
    title: str
    status: str
    content: str | None
    word_count: int | None


class SectionWrittenResponse(BaseModel):
    node_id: str
    status: str
    written_at: datetime


class SectionsResetWrittenResponse(BaseModel):
    """移除全部 AI 内容后服务端章节状态回退（决策 31）。"""

    reset_count: int


class WizardSectionLatestResponse(BaseModel):
    """跨任务取每个 node 的最新章节行（「重新写入 Word」批量动作的权威清单）。"""

    task_id: str
    node_id: str
    title: str
    status: str
    word_count: int | None

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/api/bidWizard.ts
import { getAccessToken } from "@/api/client";

const API_BASE = (import.meta.env.VITE_API_BASE || "/api").replace(/\/$/, "");

/** AI编标（bid-wizard）四阶段向导 API（doc/workspace/20-bid-wizard.md §5.2）。 */

export type WizardStage = "material" | "requirement" | "outline" | "writing";

export interface WizardSpecChart {
  type: "table" | "mermaid";
  title: string;
  points?: string | null;
}

export interface WizardSpecNode {
  node_id: string;
  title: string;
  level: number;
  summary?: string | null;
  article_count?: number;
  text_count?: number;
  charts?: WizardSpecChart[] | null;
}

export interface WizardQuestion {
  id: string;
  topic: string;
  question: string;
  why?: string;
  suggested_answer?: string;
  source?: string;
  inferred?: boolean;
  action?: "answered" | "adopted" | "skipped" | "supplemented" | null;
  answer?: string | null;
  effective_answer?: string | null;
  round?: number | null; // ≥2 为追问轮（首轮无此字段视为 1）
}

export interface Wizard {
  id: string;
  project_id: string;
  stage: WizardStage;
  status: string;
  analysis?: Record<string, unknown> | null;
  questionnaire?: {
    questions?: WizardQuestion[];
    followup?: { auto_rounds?: number; status?: "active" | "done" } | null;
  } | null;
  requirements?: { questions?: WizardQuestion[] } | null;
  spec?: WizardSpecNode[] | null;
  spec_previous?: WizardSpecNode[] | null;
  spec_confirmed_at?: string | null;
  requirements_stale: boolean;
  spec_stale: boolean;
  error_message?: string | null;
  created_at: string;
  updated_at: string;
}

export interface WizardMaterial {
  id: string;
  document_id: string;
  category?: string | null;
  index_status: "pending" | "indexing" | "indexed" | "failed";
  indexed_at?: string | null;
  index_error?: string | null;
  chunk_count?: number | null;
  original_filename?: string | null;
  doc_status?: string | null;
  word_count?: number | null;
  page_count?: number | null;
  created_at: string;
}

export interface WizardWritingTask {
  id: string;
  wizard_id: string;
  project_id: string;
  status: string;
  selected_nodes?: string[] | null;
  summary?: Record<string, unknown> | null;
  continue_of?: string | null;
  error_message?: string | null;
  started_at?: string | null;
  completed_at?: string | null;
  created_at: string;
}

export interface WizardSection {
  node_id: string;
  title: string;
  summary?: string | null;
  chart_plan?: WizardSpecChart[] | null;
  status: "pending" | "generating" | "generated" | "written" | "failed";
  word_count?: number | null;
  written_at?: string | null;
  attempts: number;
  error_message?: string | null;
}

export interface WizardSectionContent {
  node_id: string;
  title: string;
  status: string;
  content: string | null;
  word_count: number | null;
}

async function request<T>(
  method: "get" | "post" | "put" | "delete",
  path: string,
  body?: unknown,
): Promise<T> {
  const token = getAccessToken();
  const response = await fetch(`${API_BASE}${path}`, {
    method,
    headers: {
      ...(body !== undefined ? { "Content-Type": "application/json" } : {}),
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: body !== undefined ? JSON.stringify(body) : undefined,
  });
  if (!response.ok) {
    let detail: unknown = null;
    try {
      detail = (await response.json())?.detail;
    } catch {
      /* ignore body parse errors */
    }
    const error = new Error(
      typeof detail === "string"
        ? detail
        : (detail && typeof detail === "object" && "message" in detail
            ? String((detail as { message?: string }).message)
            : `请求失败（${response.status}）`),
    ) as Error & { status?: number; detail?: unknown };
    error.status = response.status;
    error.detail = detail;
    throw error;
  }
  return (await response.json()) as T;
}

export async function getWizardAccess() {
  return request<{ enabled: boolean; mode: string }>("get", "/bid-wizard/access");
}

export async function estimateIndexCost(bytes: number) {
  return request<{ chars: number; estimated_tokens: number; estimated_points: number | null }>(
    "get",
    `/bid-wizard/estimate?bytes=${encodeURIComponent(bytes)}`,
  );
}

export async function createWizard(payload: { project_id?: string; project_name?: string } = {}) {
  return request<Wizard>("post", "/bid-wizard/wizards", payload);
}

export async function getActiveWizard() {
  return request<Wizard>("get", "/bid-wizard/wizards/active");
}

export async function getWizard(wizardId: string) {
  return request<Wizard>("get", `/bid-wizard/wizards/${encodeURIComponent(wizardId)}`);
}

export async function updateWizardStage(wizardId: string, stage: WizardStage) {
  return request<Wizard>("post", `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/stage`, {
    stage,
  });
}

/** 归档当前向导（数据保留，不再被 /wizards/active 恢复）——项目列表页「归档」动作。 */
export async function archiveWizard(wizardId: string) {
  return request<Wizard>("post", `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/archive`);
}

/** 恢复归档：status 回 active、回归档前 stage 断点（决策 28，可逆动作）。 */
export async function restoreWizard(wizardId: string) {
  return request<Wizard>("post", `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/restore`);
}

/** 删除项目（决策 29）：仅已归档可删；软删 project + 异步清理 workspace；计费流水不动。 */
export async function deleteWizard(wizardId: string) {
  return request<Wizard>("delete", `/bid-wizard/wizards/${encodeURIComponent(wizardId)}`);
}

export interface WizardListItem {
  wizard_id: string;
  project_id: string;
  project_name: string;
  stage: string;
  status: string;
  tender_filename: string | null;
  latest_writing_task: { id: string; status: string } | null;
  updated_at: string;
  created_at: string;
}

/** 项目列表页聚合（§4.0）：进行中 + 已归档两组，updated_at 倒序。 */
export async function listWizards() {
  return request<{ active: WizardListItem[]; archived: WizardListItem[] }>(
    "get",
    "/bid-wizard/wizards",
  );
}

/** multipart 上传（招标文件/素材共用；走 XHR 以便带上进度回调）。 */
function uploadFile(
  path: string,
  file: File,
  extraQuery: Record<string, string> = {},
  onProgress?: (percent: number) => void,
): Promise<WizardMaterial | Wizard> {
  return new Promise((resolve, reject) => {
    const xhr = new XMLHttpRequest();
    const formData = new FormData();
    formData.append("file", file);
    const query = new URLSearchParams(extraQuery).toString();
    xhr.open("POST", `${API_BASE}${path}${query ? `?${query}` : ""}`);
    const token = getAccessToken();
    if (token) xhr.setRequestHeader("Authorization", `Bearer ${token}`);
    xhr.upload.onprogress = (event) => {
      if (event.lengthComputable && onProgress) {
        onProgress(Math.round((event.loaded / event.total) * 100));
      }
    };
    xhr.onload = () => {
      if (xhr.status >= 200 && xhr.status < 300) {
        try {
          resolve(JSON.parse(xhr.responseText));
        } catch (error) {
          reject(error);
        }
      } else {
        let detail = "";
        try {
          const parsed = JSON.parse(xhr.responseText)?.detail;
          detail = typeof parsed === "string" ? parsed : parsed?.message || `上传失败（${xhr.status}）`;
        } catch {
          detail = `上传失败（${xhr.status}）`;
        }
        const error = new Error(detail) as Error & { status?: number };
        error.status = xhr.status;
        reject(error);
      }
    };
    xhr.onerror = () => reject(new Error("网络错误，上传失败"));
    xhr.send(formData);
  });
}

export function uploadTender(wizardId: string, file: File, onProgress?: (p: number) => void) {
  return uploadFile(`/bid-wizard/wizards/${encodeURIComponent(wizardId)}/tender`, file, {}, onProgress);
}

export function uploadMaterial(
  wizardId: string,
  file: File,
  category: string | null,
  onProgress?: (p: number) => void,
) {
  return uploadFile(
    `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/materials`,
    file,
    category ? { category } : {},
    onProgress,
  );
}

export async function deleteTender(wizardId: string) {
  return request<Wizard>("delete", `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/tender`);
}

/** 删除一份招标文件（多文件口径：正文 + 补遗/澄清可并存，逐份删除）。 */
export async function deleteTenderDocument(wizardId: string, documentId: string) {
  return request<Wizard>(
    "delete",
    `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/tender/${encodeURIComponent(documentId)}`,
  );
}

/** AI 解读招标文件（同步微任务）：招标要素 + suggested_materials 建议补素材。 */
export async function analyzeTenderDocument(wizardId: string) {
  return request<Wizard>(
    "post",
    `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/analysis`,
  );
}

export async function listMaterials(wizardId: string) {
  return request<WizardMaterial[]>(
    "get",
    `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/materials`,
  );
}

export async function deleteMaterial(wizardId: string, documentId: string) {
  return request<Wizard>(
    "delete",
    `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/materials/${encodeURIComponent(documentId)}`,
  );
}

export async function getMaterialIndex(wizardId: string, documentId: string) {
  return request<{
    material_id: string;
    index_status: string;
    chunk_count: number | null;
    index_content: string | null;
  }>(
    "get",
    `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/materials/${encodeURIComponent(documentId)}/index`,
  );
}

export async function reindexMaterial(wizardId: string, documentId: string) {
  return request<WizardMaterial>(
    "post",
    `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/materials/${encodeURIComponent(documentId)}/reindex`,
  );
}

export async function generateQuestionnaire(wizardId: string) {
  return request<Wizard>(
    "post",
    `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/questionnaire`,
  );
}

/** 再次检查（反馈⑭）：保留现有问答，基于最新素材索引追加一轮 grill-me 追问。 */
export async function generateQuestionnaireRound(
  wizardId: string,
  trigger: "manual" | "auto" = "manual",
) {
  return request<Wizard>(
    "post",
    `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/questionnaire/round?trigger=${trigger}`,
  );
}

export async function saveRequirements(
  wizardId: string,
  answers: { question_id: string; action: "answered" | "adopted" | "skipped" | "supplemented"; answer?: string | null }[],
) {
  return request<Wizard>(
    "put",
    `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/requirements`,
    { answers },
  );
}

export async function generateSpec(wizardId: string) {
  return request<Wizard>("post", `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/spec`);
}

export async function saveSpec(wizardId: string, spec: WizardSpecNode[]) {
  return request<Wizard>("put", `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/spec`, {
    spec,
  });
}

export async function reviseSpec(wizardId: string, instruction: string) {
  return request<Wizard>(
    "post",
    `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/spec/revise`,
    { instruction },
  );
}

export async function rollbackSpec(wizardId: string) {
  return request<Wizard>("post", `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/spec/rollback`);
}

export async function confirmSpec(wizardId: string) {
  return request<Wizard>(
    "post",
    `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/spec/confirm`,
  );
}

export async function createWritingTask(wizardId: string, nodeIds: string[]) {
  return request<WizardWritingTask>(
    "post",
    `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/writing-tasks`,
    { node_ids: nodeIds },
  );
}

export async function getLatestWritingTask(wizardId?: string) {
  const query = wizardId ? `?wizard_id=${encodeURIComponent(wizardId)}` : "";
  return request<WizardWritingTask>("get", `/bid-wizard/writing-tasks/latest${query}`);
}

export async function getWritingTask(taskId: string) {
  return request<WizardWritingTask>("get", `/bid-wizard/writing-tasks/${encodeURIComponent(taskId)}`);
}

export async function listWritingSections(taskId: string) {
  return request<WizardSection[]>(
    "get",
    `/bid-wizard/writing-tasks/${encodeURIComponent(taskId)}/sections`,
  );
}

export async function getWritingSectionContent(taskId: string, nodeId: string) {
  return request<WizardSectionContent>(
    "get",
    `/bid-wizard/writing-tasks/${encodeURIComponent(taskId)}/sections/${encodeURIComponent(nodeId)}`,
  );
}

export async function markSectionWritten(taskId: string, nodeId: string) {
  return request<{ node_id: string; status: string; written_at: string }>(
    "post",
    `/bid-wizard/writing-tasks/${encodeURIComponent(taskId)}/sections/${encodeURIComponent(nodeId)}/written`,
  );
}

export async function regenerateWritingSection(taskId: string, nodeId: string) {
  return request<WizardWritingTask>(
    "post",
    `/bid-wizard/writing-tasks/${encodeURIComponent(taskId)}/sections/${encodeURIComponent(nodeId)}/regenerate`,
  );
}

export async function cancelWritingTask(taskId: string) {
  return request<WizardWritingTask>(
    "post",
    `/bid-wizard/writing-tasks/${encodeURIComponent(taskId)}/cancel`,
  );
}

export function wizardStreamUrl(taskId: string) {
  return `${API_BASE}/bid-wizard/writing-tasks/${encodeURIComponent(taskId)}/stream`;
}

/** 客户端素材上传端点完整地址（插件桥转发用，§5.6 material.upload）。
 * 由插件 C# HttpClient 直连 POST，必须是绝对地址：生产构建 API_BASE 为相对 `/api`，
 * 按页面 origin 补全；端点与页面本地上传同一条 POST /wizards/{id}/materials。 */
export function materialUploadUrl(wizardId: string) {
  const path = `${API_BASE}/bid-wizard/wizards/${encodeURIComponent(wizardId)}/materials`;
  return new URL(path, window.location.origin).toString();
}

/** 移除全部 AI 内容后的服务端状态回退（决策 31）：written → generated。 */
export async function resetWrittenSections(wizardId: string) {
  return request<{ reset_count: number }>(
    "post",
    `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/sections/reset-written`,
  );
}

/** 跨任务取每个 node 的最新章节行（「重新写入 Word」批量动作的权威清单）。 */
export async function listLatestSections(wizardId: string) {
  return request<
    Array<{ task_id: string; node_id: string; title: string; status: string; word_count: number | null }>
  >("get", `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/sections/latest`);
}

// ---- 多轮追问（决策 32：追问侧栏；主动追问统一到问卷轮机制，反馈⑱）----

/** 追问侧栏提问（每轮一次 bid_wizard_qa 计费）。 */
export async function askSidebarQuestion(wizardId: string, question: string) {
  return request<{ answer: string }>(
    "post",
    `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/qa/ask`,
    { question },
  );
}

/** 采纳追问问答对并入编写需求 supplementals（不计费）。 */
export async function adoptSidebarAnswer(wizardId: string, question: string, answer: string) {
  return request<Wizard>(
    "post",
    `/bid-wizard/wizards/${encodeURIComponent(wizardId)}/qa/adopt`,
    { question, answer },
  );
}

export function wizardToken() {
  return getAccessToken();
}
